src/rewards/community_advisors/mod.rs

```diff
@@ -131,38 +142,54 @@ fn load_tickets_from_reviews(
         };
     }
 
-    let excellent_reviews = proposal_reviews
-        .iter()
-        .filter(|rev| matches!(rev.score(), ReviewScore::Excellent))
-        .count() as u64;
-    let good_reviews = proposal_reviews
-        .iter()
-        .filter(|rev| matches!(rev.score(), ReviewScore::Good))
-        .count() as u64;
-
-    // assumes only one review per assessor in a single proposal
-    let ticket_distribution = proposal_reviews
-        .iter()
-        .map(|rev| {
-            let tickets = match rev.score() {
-                ReviewScore::Excellent => rewards_slots.excellent_slots,
-                ReviewScore::Good => rewards_slots.good_slots,
-                _ => unreachable!("we've already filtered out other review scores"),
-            };
-
-            (rev.assessor.clone(), tickets)
-        })
-        .collect();
-
-    let excellent_winning_tkts =
-        excellent_reviews.min(rewards_slots.max_excellent_reviews) * rewards_slots.excellent_slots;
-    let good_winning_tkts =
-        good_reviews.min(rewards_slots.max_good_reviews) * rewards_slots.good_slots;
+    // assuming only one review per assessor in a single proposal
+    let (excellent_tkts, good_tkts): (TicketsDistribution, TicketsDistribution) =
+        // a full match is used so that we don't forget to consider new review types which may be added in the future
+        proposal_reviews.iter().map(|rev| match rev.score() {
+            ReviewScore::Excellent => (rev.assessor.clone(), rewards_slots.excellent_slots),
+            ReviewScore::Good => (rev.assessor.clone(), rewards_slots.good_slots),
+            _ => unreachable!("we've already filtered out other review scores"),
+        }).partition(|(_ca, tkts)| *tkts == rewards_slots.excellent_slots);
+
+    let excellent_winning_tkts = std::cmp::min(
+        excellent_tkts.len() as u64,
+        rewards_slots.max_excellent_reviews,
+    ) * rewards_slots.excellent_slots;
+    let good_winning_tkts = std::cmp::min(good_tkts.len() as u64, rewards_slots.max_good_reviews)
+        * rewards_slots.good_slots;
+
+    ProposalTickets::Fund7 {
+        excellent_winning_tkts,
+        good_winning_tkts,
+        excellent_tkts,
+        good_tkts,
+    }
+}
 
-    ProposalTickets::Fund6 {
-        ticket_distribution,
-        winning_tkts: excellent_winning_tkts + good_winning_tkts,
+// Run a two stage lottery to reward community advisors
+//
+// In the first round, only excellent reviews will be taken into consideration
+// In the second, losing tickets from the first round will compete with good review
+fn double_lottery<R: Rng>(
+    stage1: TicketsDistribution,
+    mut stage2: TicketsDistribution,
+    to_distribute1: u64,
+    to_distribute2: u64,
```

Comment:
nitpick: I do not clearly see the meaning on `to_distribute1/2`. Maybe `distribute_first_round`, `distribute_second_round`?